virtool/samples/api.py

```diff
@@ -485,12 +486,27 @@ async def remove(req):
 
     """
     db = req.app["db"]
+    client = req["client"]
 
     sample_id = req.match_info["sample_id"]
 
     try:
-        if not await virtool.samples.db.check_rights(db, sample_id, req["client"]):
-            return insufficient_rights()
+        if isinstance(client, JobClient):
+            ready = await virtool.db.utils.get_one_field(db.samples, "ready", sample_id)
+
+            if ready is None:
+                raise virtool.errors.DatabaseError("Sample does not exist")
+
+            if ready is True:
+                return bad_request("Only unfinalized samples can be deleted")
+
+            file_ids = await virtool.db.utils.get_one_field(db.samples, "files", sample_id)
+
+            if file_ids:
+                await virtool.files.db.release(db, file_ids)
+        else:
+            if not await virtool.samples.db.check_rights(db, sample_id, client):
+                return insufficient_rights()
     except virtool.errors.DatabaseError as err:
         if "Sample does not exist" in str(err):
             return not_found()
```

Comment:
I think the best approach is to split this into two handlers: one for jobs and one for normal clients. Using conditionals is too flaky and will get very messy in some other handlers.